cpp/src/slice2java/JavaUtil.cpp

```diff
@@ -527,71 +527,89 @@ namespace
+string
+Slice::getSerialVersionUID(const ContainedPtr& p)
 {
-    ostringstream os;
-    os << "Name: " << p->scoped();
+    optional<std::int64_t> serialVersionUID = nullopt;
 
-    os << " Base: [";
-    if (p->base())
+    // Check if the user provided their own UID value with metadata.
+    string metadata;
+    if (p->findMetaData("java:serialVersionUID", metadata))
     {
-        os << p->base()->scoped();
+        string::size_type pos = metadata.rfind(":") + 1;
+        if (pos == string::npos)
+        {
+            ostringstream os;
+            os << "missing serialVersionUID value for " << p->kindOf() << " `" << p->scoped()
+               << "'; generating default value";
+            const DefinitionContextPtr dc = p->unit()->findDefinitionContext(p->file());
+            dc->warning(InvalidMetaData, "", "", os.str());
+        }
+        else
+        {
+            try
+            {
+                metadata = metadata.substr(pos);
+                serialVersionUID = std::stoll(metadata, nullptr, 0);
+            }
+            catch (const std::exception&)
+            {
+                ostringstream os;
+                os << "ignoring invalid serialVersionUID for " << p->kindOf() << " `" << p->scoped()
+                << "'; generating default value";
+                const DefinitionContextPtr dc = p->unit()->findDefinitionContext(p->file());
+                dc->warning(InvalidMetaData, "", "", os.str());
+            }
+        }
     }
-    os << "]";
 
-    os << " Members: [";
-    DataMemberList members = p->dataMembers();
-    for (DataMemberList::const_iterator i = members.begin(); i != members.end();)
+    // If the user didn't specify a UID through metadata (or it was malformed), compute a default UID instead.
+    if (!serialVersionUID)
     {
-        os << (*i)->name() << ":" << (*i)->type();
-        i++;
-        if (i != members.end())
-        {
-            os << ", ";
-        }
+        serialVersionUID = computeDefaultSerialVersionUID(p);
     }
-    os << "]";
 
-    const string data = os.str();
-    long hashCode = 5381;
-    hashAdd(hashCode, data);
-    return hashCode;
+    ostringstream os;
+    os << "private static final long serialVersionUID = " << *serialVersionUID << "L;";
+    return os.str();
 }
 
 long
-Slice::computeSerialVersionUUID(const StructPtr& p)
+Slice::computeDefaultSerialVersionUID(const ContainedPtr& p)
 {
-    ostringstream os;
-
-    os << "Name: " << p->scoped();
-    os << " Members: [";
-    DataMemberList members = p->dataMembers();
-    for (DataMemberList::const_iterator i = members.begin(); i != members.end();)
+    string name;
+    DataMemberList members;
+    optional<string> baseName;
+    if (ClassDefPtr cl = dynamic_pointer_cast<ClassDef>(p))
     {
-        os << (*i)->name() << ":" << (*i)->type();
-        i++;
-        if (i != members.end())
-        {
-            os << ", ";
-        }
+        name = cl->scoped();
```

Comment:
Can we do this outside the if blocks, for example call Contained::scoped for all of them?

Reply:
We could, but I'm actually using this to ensure that the function was called on a correct type. At the bottom is:
```
    // Assert that this was called on either a class, exception, or struct.
    assert(!name.empty());
```
If none of the casts succeed, then the name won't be assigned, and we'll detect it.

----

Now that you've pointed it out, I think I'll just delete the assert and implement your suggestion.
It's not really a big deal if this is called on other type...

Reply:
Fixed.